### PR TITLE
Ipv6 support

### PR DESCRIPTION
**Pull request by brunopeter**
- added support for ipv6 network and service groups including ipv6 addresses and masks

### CODE REVIEW

**Comment by mjbear** [comment]
@brunopeter 
This PR needs test data for the template changes that have been made.
Add another file of raw test data (which contains IPv6 objects) and its parsed equivalent.

**Comment by brunopeter** [comment]
added raw test data and parsed

**Comment by mjbear** [comment]
> added raw test data and parsed

Thank you.

Functionally, is there an advantage to separate ICMP and ICMPv6 types?

**Comment by brunopeter** [comment]
@mjbear no functional reason beyond the number of types is large and separating them made some of protocol differences more clear.

**Comment by mjbear** [comment]
> @mjbear no functional reason beyond the number of types is large and separating them made some of protocol differences more clear.

Makes sense. Thank you for replying.
I wonder if the template even needs the specific ICMP type names or if that could be a simpler regex pattern instead.

I may have to give that a try to see if it breaks anything, haha.
Of course it's also up for input from the project maintainers too. :grin:

**Comment by brunopeter** [comment]
> > @mjbear no functional reason beyond the number of types is large and separating them made some of protocol differences more clear.
> 
> 
> 
> Makes sense. Thank you for replying.
> 
> I wonder if the template even needs the specific ICMP type names or if that could be a simpler regex pattern instead.
> 
> 
> 
> I may have to give that a try to see if it breaks anything, haha.
> 
> Of course it's also up for input from the project maintainers too. :grin:

If it's like the service objects, cisco will replace numbers with the name e.g. service 80 becomes service http.

File: tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.raw [hunk 1]
```diff
@@ -0,0 +1,9 @@
+V6-Network object group TEST-v6-obj
+ host 2001:db8::1:1111
+ host 2001:db8::2:2222
+ 2001:db8:111:1::/48
+ 2001:db8:222:2::/48
+V6-Service object group TEST-v6-icmp
+ icmp echo-request
+ icmp echo-reply
+
```
Comment (jmcgill298) [inline tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.raw:9]:
```suggestion

```

File: tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    NAME: TEST-v6-obj
+    NESTED_GROUPS: ''
+    NETMASK: /48
+    NETWORK: '2001:DB8:222:2::'
+    PORT: ''
+    PORT_MATCH: ''
+    PORT_RANGE_END: ''
+    PORT_RANGE_START: ''
+    PROTOCOL: ''
+    TYPE: V6-Network
```
Comment (jmcgill298) [inline tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml:138]:
```suggestion
    TYPE: V6-Network

```

**Review by jmcgill298 — CHANGES_REQUESTED** [review]
Thanks @brunopeter just a couple of minor whitespace fixes, and will need to update the YAML file to pass linting. There are some helper scripts to fix it up for you: https://ntc-templates.readthedocs.io/en/latest/dev/dev_parser/#development-helper-scripts

File: tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    NAME: TEST-v6-obj
+    NESTED_GROUPS: ""
+    NETMASK: /48
+    NETWORK: "2001:DB8:222:2::"
+    PORT: ""
+    PORT_MATCH: ""
+    PORT_RANGE_END: ""
+    PORT_RANGE_START: ""
+    PROTOCOL: ""
+    TYPE: V6-Network
```
Comment (mjbear) [inline tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml:138]:
```suggestion
    NAME: "TEST-v6-icmp"
    NESTED_GROUPS: ""
    NETMASK: ""
    NETWORK: ""
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Service"
  - ANY: ""
    DESCRIPTION: ""
    HOST: ""
    HOST_RANGE_END: ""
    HOST_RANGE_START: ""
    ICMP6_TYPE: "echo-request"
    ICMP_TYPE: ""
    NAME: "TEST-v6-icmp"
    NESTED_GROUPS: ""
    NETMASK: ""
    NETWORK: ""
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Service"
  - ANY: ""
    DESCRIPTION: ""
    HOST: ""
    HOST_RANGE_END: ""
    HOST_RANGE_START: ""
    ICMP6_TYPE: ""
    ICMP_TYPE: "echo-reply"
    NAME: "TEST-v6-icmp"
    NESTED_GROUPS: ""
    NETMASK: ""
    NETWORK: ""
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Service"
  - ANY: ""
    DESCRIPTION: ""
    HOST: ""
    HOST_RANGE_END: ""
    HOST_RANGE_START: ""
    ICMP6_TYPE: ""
    ICMP_TYPE: ""
    NAME: "TEST-v6-obj"
    NESTED_GROUPS: ""
    NETMASK: ""
    NETWORK: ""
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Network"
  - ANY: ""
    DESCRIPTION: ""
    HOST: "2001:DB8::1:1111"
    HOST_RANGE_END: ""
    HOST_RANGE_START: ""
    ICMP6_TYPE: ""
    ICMP_TYPE: ""
    NAME: "TEST-v6-obj"
    NESTED_GROUPS: ""
    NETMASK: ""
    NETWORK: ""
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Network"
  - ANY: ""
    DESCRIPTION: ""
    HOST: "2001:DB8::2:2222"
    HOST_RANGE_END: ""
    HOST_RANGE_START: ""
    ICMP6_TYPE: ""
    ICMP_TYPE: ""
    NAME: "TEST-v6-obj"
    NESTED_GROUPS: ""
    NETMASK: ""
    NETWORK: ""
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Network"
  - ANY: ""
    DESCRIPTION: ""
    HOST: ""
    HOST_RANGE_END: ""
    HOST_RANGE_START: ""
    ICMP6_TYPE: ""
    ICMP_TYPE: ""
    NAME: "TEST-v6-obj"
    NESTED_GROUPS: ""
    NETMASK: "/48"
    NETWORK: "2001:DB8:111:1::"
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Network"
  - ANY: ""
    DESCRIPTION: ""
    HOST: ""
    HOST_RANGE_END: ""
    HOST_RANGE_START: ""
    ICMP6_TYPE: ""
    ICMP_TYPE: ""
    NAME: "TEST-v6-obj"
    NESTED_GROUPS: ""
    NETMASK: "/48"
    NETWORK: "2001:DB8:222:2::"
    PORT: ""
    PORT_MATCH: ""
    PORT_RANGE_END: ""
    PORT_RANGE_START: ""
    PROTOCOL: ""
    TYPE: "V6-Network"
```

File: ntc_templates/templates/cisco_ios_show_object-group.textfsm [hunk 4]
```diff
@@ -1,31 +1,31 @@
...
 Value NESTED_GROUPS (\S+)
 Value ANY (any)
-Value HOST (\d+\.\d+\.\d+\.\d+)
+Value HOST (((\d+\.\d+\.\d+\.\d+)|([A-Fa-f0-9:]+:+)+[A-Fa-f0-9]+))
 Value HOST_RANGE_START (\d+\.\d+\.\d+\.\d+)
 Value HOST_RANGE_END (\d+\.\d+\.\d+\.\d+)
-Value NETWORK (\d+\.\d+\.\d+\.\d+)
-Value NETMASK (\d+\.\d+\.\d+\.\d+)
+Value NETWORK ((\d+\.\d+\.\d+\.\d+)|([A-Fa-f0-9:]+:+[A-Fa-f0-9:]+))
+Value NETMASK ((\d+\.\d+\.\d+\.\d+)|((?:\/\d{1,3})))
```
Comment (mjbear) [inline ntc_templates/templates/cisco_ios_show_object-group.textfsm:10]: Remove the leading forward slash as there's little value in capturing it.

```suggestion
Value NETMASK ((\d+\.\d+\.\d+\.\d+)|((?:\d{1,3})))
```

File: ntc_templates/templates/cisco_ios_show_object-group.textfsm [hunk 5]
```diff
@@ -1,31 +1,31 @@
...
 
 Start
-  ^(Service|Network) -> Continue.Clearall
+  ^(Service|Network|V6-Service|V6-Network) -> Continue.Clearall
   ^${TYPE}\s+object\s+group\s+${NAME}\s*$$ -> Record
   ^\s+Description\s+${DESCRIPTION}$$ -> Record
   ^\s+group-object\s+${NESTED_GROUPS}\s*$$ -> Record
-  ^\s+(host\s+${HOST}|range\s+${HOST_RANGE_START}\s+${HOST_RANGE_END}|${ANY}|${NETWORK}\s+${NETMASK})\s*$$ -> Record
-  ^\s+icmp\s+${ICMP_TYPE}\s*$$ -> Record
+  ^\s+(host\s+${HOST}|range\s+${HOST_RANGE_START}\s+${HOST_RANGE_END}|${ANY}|${NETWORK}\s*${NETMASK})\s*$$ -> Record
```
Comment (mjbear) [inline ntc_templates/templates/cisco_ios_show_object-group.textfsm:24]:
```suggestion
  ^\s+(host\s+${HOST}|range\s+${HOST_RANGE_START}\s+${HOST_RANGE_END}|${ANY}|${NETWORK}\s*/${NETMASK})\s*$$ -> Record
```

File: tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml (hunk 3 above)
Comment (mjbear) [reply tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml:138]: @brunopeter
I didn't realize this until now, but the keys in your YAML are in uppercase but they should be lowercase.

I'm going to work up a PR against your ipv6-support branch that you can merge in to fix this up. Thank you!

Comment (mjbear) [reply tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml:138]: Canceling this inline review as they're obsoleted by brunopeter/ntc-templates/pull/1
Please review that PR and merge it in. Thank you!

File: ntc_templates/templates/cisco_ios_show_object-group.textfsm (hunk 4 above)
Comment (mjbear) [reply ntc_templates/templates/cisco_ios_show_object-group.textfsm:10]: Obsoleted by brunopeter/ntc-templates/pull/1

File: ntc_templates/templates/cisco_ios_show_object-group.textfsm (hunk 5 above)
Comment (mjbear) [reply ntc_templates/templates/cisco_ios_show_object-group.textfsm:24]: Obsoleted by brunopeter/ntc-templates/pull/1

File: tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml (hunk 2 above)
Comment (mjbear) [reply tests/cisco_ios/show_object-group/cisco_ios_show_object-group-v6.yml:138]: This will be obsoleted by brunopeter/ntc-templates/pull/1

**Comment by mjbear** [comment]
@brunopeter
Lots going on here so my apologies for the flurry of things.

I've pulled together a PR against your fork of ntc-templates so you can merge some changes into your ipv6-support branch. **Please review that PR and merge it into your ipv6-branch once you've read through it and follow the changes.**

Any questions or comments, put them on the discussion for the PR brunopeter/ntc-templates/pull/1 over at your forked repo.

**Review by jmcgill298 — CHANGES_REQUESTED** [review]

**Comment by mjbear** [comment]
> @brunopeter Lots going on here so my apologies for the flurry of things.
> 
> I've pulled together a PR against your fork of ntc-templates so you can merge some changes into your ipv6-support branch. **Please review that PR and merge it into your ipv6-branch once you've read through it and follow the changes.**
> 
> Any questions or comments, put them on the discussion for the PR [brunopeter/pull/1](https://github.com/brunopeter/ntc-templates/pull/1) over at your forked repo.

@brunopeter
Have you seen this?

**Comment by mjbear** [comment]
> > @brunopeter Lots going on here so my apologies for the flurry of things.
> > I've pulled together a PR against your fork of ntc-templates so you can merge some changes into your ipv6-support branch. **Please review that PR and merge it into your ipv6-branch once you've read through it and follow the changes.**
> > Any questions or comments, put them on the discussion for the PR [brunopeter/pull/1](https://github.com/brunopeter/ntc-templates/pull/1) over at your forked repo.
> 
> @brunopeter Have you seen this?

Hi @brunopeter
Could you take a look at this when you have time?
Thank you!

**Comment by mjbear** [comment]
@brunopeter 
Did you mean to close this PR?

I was being nice. :disappointed:

**Pull request closed without merging**